File: softeer/로드밸런서트래픽예측.java
import java.util.*;
import java.io.*;

class Main{
    static int n;
    static long k;
    static int indegree[];
    static ArrayList<Integer> child[];
    static long result[];
    public static void main(String args[]) throws Exception{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        String inp[] = br.readLine().split(" ");

        n = Integer.parseInt(inp[0]);
        k = Long.parseLong(inp[1]);

        indegree = new int[n+1];
        result = new long[n+1];
        child = new ArrayList[n+1];

        for(int i = 1; i <= n; i++){
            child[i] = new ArrayList<Integer>();
            inp = br.readLine().split(" ");

            int p = Integer.parseInt(inp[0]);

            for(int j = 1; j <= p; j++){
                int nw = Integer.parseInt(inp[j]);
                indegree[nw]++;
                child[i].add(nw);
            }
        }

        findResult(1,k);

        for(int i = 1; i<= n; i++){
            bw.append(result[i] + " ");
        }
        bw.append("\n");
        bw.flush();
        br.close();
        bw.close();
    }

    static void findResult(int idx,long val){
        result[idx] += val;
        
        if(indegree[idx] > 1){
            indegree[idx]--;
            return;
        }
        
        long childNum = child[idx].size();

        if(childNum > 0){
            long nVal = result[idx] / childNum;
            long left = result[idx] % childNum;

            for(int i = 0; i < childNum; i++){
                if(left > 0){
                    left--;
                    findResult(child[idx].get(i),nVal + 1);
                }else{
                    findResult(child[idx].get(i),nVal);
                }
            }
        }
    }
}

File: softeer/로드밸런서트래픽예측.cpp
#include <iostream>
#include <vector>

using namespace std;

long long n,k;

vector<int> node[100001];
int childIdx[100001];
int indegree[100001];
long long result[100001];

void findNext(int idx, long long k){
    result[idx] = result[idx] + k;

    int childNum = node[idx].size();

    if(indegree[idx] > 1){
        indegree[idx]--;
        return;
    }

    if(childNum > 0){
        long long nK = result[idx] / childNum;
        long long nleft = result[idx] % childNum;

        for(int i = 0; i < childNum; i++){
            if(nleft > 0){
                findNext(node[idx][i],nK + 1);
                nleft--;
            }else{
                findNext(node[idx][i],nK);
            }
        }
    }
}

int main(){
    cin >> n >> k;

    for(int i = 1; i <= n; i++){
        int p;
        cin >> p;
        for(int j = 0; j < p; j++){
            int nextWorker;
            cin >> nextWorker;
            indegree[nextWorker]++;
            node[i].push_back(nextWorker);
        }
    }
    
    findNext(1,k);


    for(int i = 1; i <= n; i++){
        cout << result[i] << " ";
    }
    cout << "\n";

    return 0;
}